pke/kyber/kyber768/internal/cpapke.go

```diff
@@ -172,5 +174,5 @@ func (sk *PrivateKey) Equal(other *PrivateKey) bool {
 			ret |= sk.sh[i][j] ^ other.sh[i][j]
 		}
 	}
-	return ret == 0
+	return subtle.ConstantTimeEq(int32(ret), 0) == 1
```

Comment:
This doesn't do anything: the code generated is the same. https://godbolt.org/z/eGfrdcr7s

ConstantTimeEq exists so you can pass its value to a constant-time select or copy. In this case, the (in)equality of the _whole of_ the private key is not sensitive, so the current code is fine.

Reply:
`subtle.ConstantTimeEq(v, 0) == 1` is already used widely throughout the library. See dilithiums `(*PrivateKey).Equal` and bls12381/ff's `ctUint64Eq`.

Reply:
To quote Dilithium's `PrivateKey.Equal()`:

```
    ret := (subtle.ConstantTimeCompare(sk.rho[:], other.rho[:]) &
        subtle.ConstantTimeCompare(sk.key[:], other.key[:]) &
        subtle.ConstantTimeCompare(sk.tr[:], other.tr[:]))

    acc := uint32(0)
    for i := 0; i < L; i++ {
        for j := 0; j < common.N; j++ {
            acc |= sk.s1[i][j] ^ other.s1[i][j]       
        }                                             
    }
    for i := 0; i < K; i++ {          
        for j := 0; j < common.N; j++ {
            acc |= sk.s2[i][j] ^ other.s2[i][j]
            acc |= sk.t0[i][j] ^ other.t0[i][j]
        }                                                                     
    }
return (ret & subtle.ConstantTimeEq(int32(acc), 0)) == 1
```

The equivalent use to your proposal would be:

```
return subtle.ConstantTimeEq(int32(acc)(ret & subtle.ConstantTimeEq(int32(acc), 0)),1) == 1)
```

Regardless, my original point stands: the code generated between the current original in Kyber and your proposal is the same.

Reply:
Huh? The equivalent code is right there: `subtle.ConstantTimeEq(int32(acc), 0)`, and the generated code looks different to me. The dilithium code could easily look like `return ret | acc != 0` which would be the same as what is currently done. Regardless I can remove it if you would like.

Reply:
The code generated is different, but in both cases it's constant time. The reason `ConstantTimeEq` exists, is to be able to pass its return value into `ConstantTimeSelect`. Indeed: regular `==` between uint32, will give you the right bool in constant time, but you can't easily convert that `bool` into the integer required for `ConstantTimeSelect`. In other languages `bool`s are also dangerous because the compiler likes to get clever, but in Go that's less of a concern.